File: java/src/test/java/org/knguyenc/mathematics/FibonacciTest.java
package org.knguyenc.mathematics;

import junit.framework.TestCase;

public class FibonacciTest extends TestCase {

    public void testVarietyInputs() {
        assertEquals(8, Fibonacci.binaryFibonacci(6));
        assertEquals(55, Fibonacci.binaryFibonacci(10));
        assertEquals(21, Fibonacci.binaryFibonacci(8));

        assertEquals(8, Fibonacci.linearFibonacci(6)[0]);
        assertEquals(55, Fibonacci.linearFibonacci(10)[0]);
        assertEquals(21, Fibonacci.linearFibonacci(8)[0]);
    }
}

File: mathematics/fibonacci_test.cc
#include "fibonacci.h"

#include <gtest/gtest.h>

namespace algorithm {
namespace {

TEST(FibonacciTest, HandlesZeroInput) {
  EXPECT_EQ(mathematics::BinaryFibonacci(0), 0);
  EXPECT_EQ(std::get<0>(mathematics::LinearFibonacci(0)), 0);
}

TEST(FibonacciTest, HandlesPositiveInput) {
  EXPECT_EQ(mathematics::BinaryFibonacci(1), 1);
  EXPECT_EQ(mathematics::BinaryFibonacci(2), 1);
  EXPECT_EQ(mathematics::BinaryFibonacci(3), 2);
  EXPECT_EQ(mathematics::BinaryFibonacci(6), 8);

  EXPECT_EQ(std::get<0>(mathematics::LinearFibonacci(1)), 1);
  EXPECT_EQ(std::get<0>(mathematics::LinearFibonacci(2)), 1);
  EXPECT_EQ(std::get<0>(mathematics::LinearFibonacci(3)), 2);
  EXPECT_EQ(std::get<0>(mathematics::LinearFibonacci(6)), 8);
}

TEST(FibonacciTest, HandlesNegativeInput) {
  EXPECT_DEATH(mathematics::BinaryFibonacci(-1), "");
  EXPECT_DEATH(mathematics::BinaryFibonacci(-100), "");

  EXPECT_DEATH(mathematics::LinearFibonacci(-1), "");
  EXPECT_DEATH(mathematics::LinearFibonacci(-100), "");
}

}  // namespace
}  // namespace algorithm
